### ISSUE TITLE
Add static analysis to find bugs earlier

### ISSUE THREAD

**Issue by mbrukman** [GoogleCloudPlatform/google-cloud-cpp#10]
There are a number of options for finding issues via static analysis, whether via self-run tooling or via SaaS, e.g.,

## Specific tools

* [CppCheck](http://cppcheck.sourceforge.net/) — open-source, self-managed
* [Coverity Scan](https://scan.coverity.com/) — commercial SaaS, free for open-source GitHub repos — most of their [success stories](https://scan.coverity.com/o/oss_success_stories) are for C/C++ repositories!
* [Clang static analyzer](https://clang-analyzer.llvm.org/) — open-source, self-managed
* [FlawFinder](https://www.dwheeler.com/flawfinder/) — a new one I found, but haven't heard much about it in the past

## Collections, lists, reviews, etc.

* [a list of static analysis tools](https://en.wikipedia.org/wiki/List_of_tools_for_static_code_analysis#C.2C_C.2B.2B) (Wikipedia)
* [another list of tools](https://github.com/mre/awesome-static-analysis#cc) (via GitHub repo)
* [yet another list of tools](https://www.peerlyst.com/posts/a-list-of-static-analysis-tools-for-c-c-peerlyst) which separates tools into 3 categories: (i) free/open-source, (ii) free for academic use, and (iii) commercial
* [a Stack Overflow discussion about C++ static analysis tools](https://stackoverflow.com/questions/141498/what-open-source-c-static-analysis-tools-are-available)
* [a review of C++ static analyzers](https://hownot2code.com/2017/06/29/review-of-contemporary-cc-static-code-analyzers/) (June 2017)
* [static code analysis comparison by John Carmack (of Quake fame)](https://www.viva64.com/en/a/0087/); posted on the PVS-Studio blog because his original article is no longer available

I think we should start with enabling Coverity Scan and integrating Clang analyzers (since we're going to use Clang for building our code anyway).

Adding CppCheck and FlawFinders are also a good idea, but if they don't support C++11 (or C++14, if/when we upgrade this repo's minimum C++ requirements), that might become an issue.

**Comment by coryan** [comment]
This is a great idea.  Some general comments:

* You get cppcheck for "free" if we use codacy.com.
* Coverity Scan is great, but a bit of pain to setup.  If anybody wants to set it up, come talk to me about gotchas.

**Comment by mbrukman** [comment]
I've experienced issues with setting up Coverity Scan on Travis (for another project) as well; we couldn't get the native Travis CI + Coverity Scan integration to work, so we ended up writing our own script + Dockerfile, and manually invoke the Coverity Scan build as another build; here's what we ended up with, which works now, but was certainly a long road to get there:

* https://github.com/JanusGraph/janusgraph/blob/master/.travis.yml (search for "Coverity"; note that we're not using native Travis CI support)
* https://github.com/JanusGraph/janusgraph/tree/master/analysis (custom script + Dockerfile)

**Comment by coryan** [comment]
tl;dr; @mbrukman I need some ideas on how to export / publish the results of clang's static analyzer.

I have a branch running the clang static analyzer via `scan-build`:

https://github.com/coryan/google-cloud-cpp/tree/enable-scan-build

And it does detect things (when I force errors for it to detect):

https://travis-ci.org/coryan/google-cloud-cpp/jobs/314187499#L6717
https://travis-ci.org/coryan/google-cloud-cpp/jobs/314187499#L6696

The results are interesting, but they need to be hosted on a web server to be reviewed.  Or at least stored in a tarball that folks can download to see locally (it is a purely static website).  Any thoughts on how we would do that?

**Comment by mbrukman** [comment]
@coryan — some options I thought of, approximately from least to most effort:

1. run it on Travis as a separate build in the matrix, but allow it to fail (such that it doesn't stop merging PRs, unless you want to force it to block PR submission), periodically (manually) review results
2. run it on Travis, and publish to a GitHub repo which is published on the web; periodically (manually) review the results
    * needs more work than just submitting a file to git to be able to visualize results, find errors, ability to link to results, etc.
    * note that Travis highlights builds that fail and makes it easy to review logs
3. find a service that provides Clang scan + result hosting like Coverity, Codacy, Codecov, etc. do for their tools (I am not aware of such a tool's existence, so this could be a fruitless search)
4. build a custom web app to host this — very involved, possibly more trouble than the value it will provide

Seems like we can get quite a lot of value from just doing option 1. Thoughts?

**Comment by coryan** [comment]
I think we should go with something like (1).  For what is worth, running scan-build on Travis pushes the build times to ~40 minutes:

**Comment by mbrukman** [comment]
**Re: caching:** Travis CI supports caching, but it has to be compatible with all builds, or we have to be careful to create per-matrix-item cache directory so that they don't conflict (since some builds might be Debug, some might be Release, they may use different flags, etc.). https://docs.travis-ci.com/user/caching/ has the info; take a look and see if you think that would work for gRPC — I think it should.

**Re: 50min timeout:** note that the 50min built timeout is for any job (a build consists of N parallel jobs); it's not the total CPU time taken; e.g., you can see that [this JanusGraph build](https://travis-ci.org/JanusGraph/janusgraph/builds/317739467) took much longer than 50min in total across all jobs, but no single job exceeded 50min on its own.

**Comment by coryan** [comment]
> Re: 50min timeout: note that the 50min built timeout is for any job (a build consists of N parallel jobs); it's 

Yup, just building *one* configuration with the clang static analyzer takes 37 minutes though.

>  https://docs.travis-ci.com/user/caching/ has the info; take a look and see if you think that would work for gRPC — I think it should.

Yup, we used to do that for the docker images, remember?  And downloading / uploading the cache took longer than rebuilding from scratch.  Anyway, not urgent.

I am going to send a PR to enable the clang static analyzer build.  If it becomes a hassle we can change it.

**Comment by coryan** [comment]
I am going to declare this closed since we enabled a Clang static analyzer build.  @mbrukman if you want to reopen to enable Coverity Scan, or some other analyzer please feel free.

**Comment by mbrukman** [comment]
[Codacy](https://www.codacy.com/) is a non-starter; they require _way_ too many permissions (emphasis theirs):

> ### Repositories
> Public only
> This application will be able to **read and write all public repository data**. This includes the following:
> 
> - Code
> - Issues
> - Pull requests
> - Wikis
> - Settings
> - Webhooks and services
> - Deploy keys

I'll think about Coverity in my copious spare time.
